### ISSUE TITLE
[Feature] Add Mixed Precision Training Support

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ProjectOdyssey#2669]
## Problem

Training with Float32 is slower and uses more memory than necessary. Mixed precision (FP16/FP32) can provide 2-3x speedup.

### 1. Automatic Mixed Precision (AMP)

```mojo
# shared/training/amp.mojo
struct AutocastContext:
    """Automatically cast operations to FP16 for performance."""
    var enabled: Bool
    var dtype: DType
    
    fn __init__(out self, enabled: Bool = True, dtype: DType = DType.float16):
        self.enabled = enabled
        self.dtype = dtype
    
    fn __enter__(mut self):
        if self.enabled:
            enable_autocast(self.dtype)
    
    fn __exit__(mut self):
        if self.enabled:
            disable_autocast()

# Usage
fn train_step(model: Model, optimizer: Optimizer):
    with AutocastContext(enabled=True, dtype=DType.float16):
        # Forward pass in FP16
        var output = model(input)
        var loss = criterion(output, target)
    
    # Backward pass (scaled for FP16 stability)
    var scaler = GradScaler()
    scaler.scale(loss).backward()
    scaler.step(optimizer)
    scaler.update()
```

### 2. Gradient Scaling

```mojo
struct GradScaler:
    """Scale gradients to prevent underflow in FP16."""
    var scale: Float32
    var growth_factor: Float32
    var backoff_factor: Float32
    var growth_interval: Int
    var _growth_tracker: Int
    
    fn __init__(out self, init_scale: Float32 = 65536.0):
        self.scale = init_scale
        self.growth_factor = 2.0
        self.backoff_factor = 0.5
        self.growth_interval = 2000
        self._growth_tracker = 0
    
    fn scale(self, loss: ExTensor) -> ExTensor:
        """Scale loss before backward."""
        return loss * self.scale
    
    fn step(mut self, optimizer: Optimizer):
        """Unscale gradients and step optimizer."""
        # Unscale gradients
        for param in optimizer.parameters():
            if param.grad:
                param.grad = param.grad / self.scale
        
        # Check for inf/nan
        var has_inf = self._check_inf_nan(optimizer.parameters())
        
        if not has_inf:
            optimizer.step()
            self._growth_tracker += 1
            
            # Increase scale if stable
            if self._growth_tracker >= self.growth_interval:
                self.scale *= self.growth_factor
                self._growth_tracker = 0
        else:
            # Decrease scale if overflow detected
            self.scale *= self.backoff_factor
            self._growth_tracker = 0
            print("Gradient overflow detected, reducing scale to", self.scale)
```

### 3. Per-Operation Precision

```mojo
# Some operations benefit from FP32 even in mixed precision
fn mixed_precision_forward(input: ExTensor) -> ExTensor:
    # Most operations in FP16
    with AutocastContext(dtype=DType.float16):
        var x = conv1(input.cast(DType.float16))
        var x = relu(x)
        var x = conv2(x)
    
    # Batch norm in FP32 for stability
    var x = batch_norm(x.cast(DType.float32))
    
    # Output in FP16
    return x.cast(DType.float16)
```

### 4. Training Loop with AMP

```mojo
fn train_with_amp():
    var model = LeNet5()
    var optimizer = Adam(model.parameters(), lr=0.001)
    var scaler = GradScaler()
    
    for epoch in range(num_epochs):
        for batch in dataloader:
            optimizer.zero_grad()
            
            # Forward in FP16
            with AutocastContext(dtype=DType.float16):
                var output = model(batch.data.cast(DType.float16))
                var loss = cross_entropy(output, batch.target)
            
            # Backward with gradient scaling
            scaler.scale(loss).backward()
            
            # Optimizer step with unscaling
            scaler.step(optimizer)
            scaler.update()
```

### 5. Memory Savings

```mojo
# Estimate memory savings
fn compare_memory_usage():
    var size = [128, 3, 224, 224]  # Batch of images
    
    # FP32: 4 bytes per element
    var fp32_size = product(size) * 4
    print("FP32 memory:", fp32_size / 1_000_000, "MB")
    
    # FP16: 2 bytes per element
    var fp16_size = product(size) * 2
    print("FP16 memory:", fp16_size / 1_000_000, "MB")
    
    print("Memory savings:", (1 - fp16_size / fp32_size) * 100, "%")
```

## Benefits

- **Speed**: 2-3x faster training on modern hardware
- **Memory**: 2x less memory usage
- **Scale**: Train larger models or bigger batches
- **Standard practice**: Used by PyTorch, TensorFlow

## Considerations

- Not all operations benefit (some need FP32 for stability)
- Requires gradient scaling to prevent underflow
- Need to test model convergence with FP16

## Impact

- **Severity**: LOW
- **Performance**: 2-3x training speedup
- **Memory**: 2x reduction in memory usage

**Comment by mvillmow** [comment]
## Implementation Update

After analyzing the codebase and attempting to implement the AutocastContext, I've encountered a key limitation:

### Mojo Limitation Discovered

Mojo v0.25.7 does not support mutable module-level global variables, which is essential for implementing PyTorch-style autocast contexts that maintain global state across nested contexts.

**What Was Attempted:**
1. AutocastContext struct with enter/exit methods
2. Global state management for nested context preservation
3. Comprehensive test suite (10 test functions)

**Limitation:**
- Cannot use `var` at module level (compiler error: "global vars are not supported")
- Cannot mutate `comptime`/`alias` values at runtime
- Global functions like `enable_autocast()` / `disable_autocast()` become no-ops

### Current Infrastructure is Sufficient

The good news: **ML Odyssey already has comprehensive mixed precision support** through:

1. **PrecisionConfig** - Unified configuration for FP32/FP16/BF16/FP8
   - Automatic gradient scaler management
   - Master weight handling  
   - Tensor casting utilities

2. **GradientScaler** - Production-ready gradient scaling
   - Dynamic scale adjustment
   - Overflow detection and backoff
   - Growth interval management

3. **Existing Usage Pattern:**
   ```mojo
   # Instead of autocast context:
   var config = PrecisionConfig.fp16()
   var input_fp16 = config.cast_to_compute(input)
   var output = model.forward(input_fp16)
   var scaled_loss = config.scale_loss(loss)
   ```

### Recommendation

**Close this issue** or mark as "waiting for Mojo mutable globals support". The proposed AutocastContext adds marginal value over existing PrecisionConfig, and the implementation is blocked by fundamental language limitations.

The existing infrastructure already provides:
- ✅ Gradient scaling
- ✅ Multiple precision modes
- ✅ Master weight management
- ✅ Type-safe precision control
- ✅ Production-ready and tested

The only missing piece (automatic context-based casting) requires Mojo language features that don't exist yet.

---

**Alternative**: If AutocastContext is still desired, implement it as a **local state wrapper** around PrecisionConfig (no global state), where users explicitly pass the context to operations. This would work but differs significantly from PyTorch's design.

**Comment by mvillmow** [comment]
Resolution: Closing - sufficient infrastructure exists. See previous implementation update comment for details.
